File: cpp_merge/packages/integration/SWEagent/tools/src/image-tools/view-image.cpp
#include "view-image.h"

any getMimeType(string filename)
{
    auto ext = path->extname(filename)->toLowerCase();
    static switch_type __switch497_790 = {
        { any(std:(".png")), 1 },
        { any(std:(".jpg")), 2 },
        { any(std:(".jpeg")), 3 },
        { any(std:(".webp")), 4 },
        { any(std:(".gif")), 5 },
        { any(std:(".svg")), 6 }
    };
    switch (__switch497_790[ext])
    {
    case 1:
        return std:("image/png");
    case 2:
    case 3:
        return std:("image/jpeg");
    case 4:
        return std:("image/webp");
    case 5:
        return std:("image/gif");
    case 6:
        return std:("image/svg+xml");
    default:
        return nullptr;
    }
};


void viewImage(string imagePath)
{
    if (!fs::existsSync(imagePath)) {
        console->error(std:("Error: File '") + imagePath + std:("' does not exist"));
        process->exit(1);
    }
    auto stats = fs::statSync(imagePath);
    if (!stats->isFile()) {
        console->error(std:("Error: '") + imagePath + std:("' is not a file"));
        process->exit(1);
    }
    auto mimeType = getMimeType(imagePath);
    if (OR((!mimeType), (!VALID_MIME_TYPES->has(mimeType)))) {
        console->error(std:("Error: Unsupported image type: ") + mimeType + std:(". Valid types are: ") + Array->from(VALID_MIME_TYPES)->join(std:(", ")) + string_empty);
        process->exit(1);
    }
    try
    {
        auto imageBuffer = fs::readFileSync(imagePath);
        auto base64 = imageBuffer->toString(std:("base64"));
        console->log(std:("![") + path->basename(imagePath) + std:("](data:") + mimeType + std:(";base64,") + base64 + std:(")"));
    }
    catch (const any& error)
    {
        console->error(std:("Error processing image: ") + error + string_empty);
        process->exit(1);
    }
};


void setupCLI()
{
    program->name(std:("view-image"))->description(std:("View an image file as a base64-encoded markdown image"))->version(std:("1.0.0"))->argument(std:("<image-file>"), std:("Path to the image file to view"))->action([=](auto imagePath) mutable
    {
        viewImage(imagePath);
    }
    );
    program->parse(process->argv);
};


std::shared_ptr<Set<string>> VALID_MIME_TYPES = std::make_shared<Set>(array<string>{ std:("image/png"), std:("image/jpeg"), std:("image/jpg"), std:("image/webp"), std:("image/gif"), std:("image/svg+xml") });

void Main(void)
{
    if (OR((require->main == module), (require->main->filename->endsWith(std:("/bin/view_image"))))) {
        setupCLI();
    }
}

MAIN

File: SWEagent/tools/src/image-tools/view-image.ts
#!/usr/bin/env node
/**
 * View image tool
 * Displays an image file as a base64-encoded markdown image
 * Converted from tools/image_tools/bin/view_image
 */

import * as fs from 'fs';
import * as path from 'path';
import { program } from 'commander';

const VALID_MIME_TYPES = new Set([
  'image/png',
  'image/jpeg', 
  'image/jpg',
  'image/webp',
  'image/gif',
  'image/svg+xml'
]);

function getMimeType(filename: string): string | null {
  const ext = path.extname(filename).toLowerCase();
  switch (ext) {
    case '.png':
      return 'image/png';
    case '.jpg':
    case '.jpeg':
      return 'image/jpeg';
    case '.webp':
      return 'image/webp';
    case '.gif':
      return 'image/gif';
    case '.svg':
      return 'image/svg+xml';
    default:
      return null;
  }
}

function viewImage(imagePath: string): void {
  // Check if file exists
  if (!fs.existsSync(imagePath)) {
    console.error(`Error: File '${imagePath}' does not exist`);
    process.exit(1);
  }

  // Check if it's a file
  const stats = fs.statSync(imagePath);
  if (!stats.isFile()) {
    console.error(`Error: '${imagePath}' is not a file`);
    process.exit(1);
  }

  // Get MIME type
  const mimeType = getMimeType(imagePath);
  if (!mimeType || !VALID_MIME_TYPES.has(mimeType)) {
    console.error(`Error: Unsupported image type: ${mimeType}. Valid types are: ${Array.from(VALID_MIME_TYPES).join(', ')}`);
    process.exit(1);
  }

  try {
    // Read file and convert to base64
    const imageBuffer = fs.readFileSync(imagePath);
    const base64 = imageBuffer.toString('base64');
    
    // Output markdown image syntax
    console.log(`![${path.basename(imagePath)}](data:${mimeType};base64,${base64})`);
  } catch (error) {
    console.error(`Error processing image: ${error}`);
    process.exit(1);
  }
}

// CLI if run directly
// CLI setup
function setupCLI() {
  program
    .name('view-image')
    .description('View an image file as a base64-encoded markdown image')
    .version('1.0.0')
    .argument('<image-file>', 'Path to the image file to view')
    .action((imagePath) => {
      viewImage(imagePath);
    });

  program.parse(process.argv);
}

// Run CLI if called directly or from bin script
if (require.main === module || require.main?.filename?.endsWith('/bin/view_image')) {
  setupCLI();
}

export { viewImage };
